src/partial-evaluators/CallExpression.js

```diff
@@ -9,16 +9,199 @@
 
 /* @flow */
 
-import type { BabelNodeCallExpression, BabelNodeStatement } from "babel-types";
+import type { BabelNodeCallExpression, BabelNodeExpression, BabelNodeStatement } from "babel-types";
 import type { Realm } from "../realm.js";
 import type { LexicalEnvironment } from "../environment.js";
 
-import { AbruptCompletion } from "../completions.js";
-import { Value } from "../values/index.js";
+import { AbruptCompletion, Completion, PossiblyNormalCompletion } from "../completions.js";
+import { TypesDomain, ValuesDomain } from "../domains/index.js";
+import { EnvironmentRecord, Reference } from "../environment.js";
+import { composeNormalCompletions, EvaluateDirectCallWithArgList,
+   GetBase, GetReferencedName, GetThisValue, GetValue,
+   IsInTailPosition, IsPropertyReference,
+   joinEffects, PerformEval,
+   SameValue, stopEffectCaptureAndJoinCompletions, unbundleNormalCompletion } from "../methods/index.js";
+import { AbstractValue, BooleanValue, FunctionValue, Value } from "../values/index.js";
 
+import * as t from "babel-types";
+import invariant from "../invariant.js";
+
+// ECMA262 12.3.4.1
 export default function (
   ast: BabelNodeCallExpression, strictCode: boolean, env: LexicalEnvironment, realm: Realm
-): [AbruptCompletion | Value, BabelNodeCallExpression, Array<BabelNodeStatement>] {
-  let result = env.evaluateCompletionDeref(ast, strictCode);
-  return [result, ast, []];
+): [Completion | Value, BabelNodeExpression, Array<BabelNodeStatement>] {
+  realm.setNextExecutionContextLocation(ast.loc);
+
+  // 1. Let ref be the result of evaluating MemberExpression.
+  let [ref, calleeAst, calleeIO] = env.partiallyEvaluateCompletion(ast.callee, strictCode);
+  if (ref instanceof AbruptCompletion)
+    return [ref, (calleeAst: any), calleeIO];
+  let completion;
+  if (ref instanceof PossiblyNormalCompletion) {
+    completion = ref;
+    ref = completion.value;
+  }
+  invariant(ref instanceof Value || ref instanceof Reference);
+
+  // 2. Let func be ? GetValue(ref).
+  let func = GetValue(realm, ref);
+
+  let io = calleeIO;
+  let partialArgs = [];
+  let argVals = [];
+  for (let arg of ast.arguments) {
+    let [argValue, argAst, argIO] = env.partiallyEvaluateCompletionDeref(arg, strictCode);
+    io = io.concat(argIO);
+    partialArgs.push((argAst: any));
+    if (argValue instanceof AbruptCompletion) {
+      if (completion instanceof PossiblyNormalCompletion)
+        completion = stopEffectCaptureAndJoinCompletions(completion, argValue, realm);
+      else
+        completion = argValue;
+      let resultAst = t.callExpression((calleeAst: any), partialArgs);
```

Comment:
Do you mean the cast to any? I am unhappy about that, but 1) there is no convenient way to add a suitable invariant (because of the large union involved) and 2) a more principled solution with the aid of generics is far too intrusive for this pull request.